site/blog/unlisted/why-sql-for-infrastructure.mdx

```diff
@@ -0,0 +1,94 @@
+
+- By defining [foreign key relations](https://www.postgresql.org/docs/current/tutorial-fk.html) between tables, the database can immediately inform you if you have dependencies that must be created before you can create the resource you desire.
+- By defining [unique, not-null, and check constraints](https://www.postgresql.org/docs/current/ddl-constraints.html) on top of a [rich type system](https://www.postgresql.org/docs/current/datatype.html) the database can prevent you from inserting faulty data to a greater degree than even Rust, as the constraints can be against the live state of your infrastructure itself.
+- By being in a database, you can answer questions about your infrastructure via queries, and in the case of security vulnerabilities or expensive misconfigurations you can update that state across the board at the same time.
+- By being a bijective relationship with your cloud, you can re-synchronize the database with your cloud at any time, avoiding statefile issues by never getting too out-of-date, and being more difficult to corrupt.
+- By defining [trigger functions](https://www.postgresql.org/docs/current/sql-createtrigger.html#SQL-CREATETRIGGER-EXAMPLES) an [audit log](https://en.wikipedia.org/wiki/Audit_trail) can be maintained, which can keep track of who made what cloud changes and when they did.
+- Other trigger functions can be added to automatically re-insert deleted records for automated [Chaos Engineering](https://en.wikipedia.org/wiki/Chaos_engineering) recovery.
+- [Snapshotting](https://en.wikipedia.org/wiki/Snapshot_%28computer_storage%29) could be used to quickly revert all infrastructure changes without hidden state (eg, the configuration of your load balancers, but not the data within your own databases).
+- By having a standardized protocol to interface with the database, you can:
+  - Perform operations on it via scripts and the [`psql` CLI](https://www.postgresql.org/docs/current/app-psql.html) tool.
+  - Integrate it into your application itself with a [postgres client library](https://node-postgres.com/) for application-driven infrastructure management (like provisioning sharded resources for a client that wants isolation).
+  - Connect to it with a [SQL IDE](https://www.jetbrains.com/datagrip/features/postgresql/) for outage mitigation or inspection in a tabular (Excel-like) view.
+  - Connect it to a [graphical dashboard](https://grafana.com/docs/grafana/latest/datasources/postgres/) for live visualization of your infrastructure.
+  - Connect it to a [low code tool](https://docs.retool.com/docs/postgresql-integration) for rapid response dashboards for your infrastructure with buttons you can tackle known outage vectors near-instantly.
+
+## SQL is the best option for Infrastructure Management
```

Comment:
I'm not sure if `Infra as data` makes more sense than `infra as SQL` in the title and this heading. I'm afraid of people thinking about trying to version control SQL - https://twitter.com/IMTantri/status/1632242097611849729.

It might be worth it to extend on this post that trended a while back https://www.reddit.com/r/kubernetes/comments/nz9fxd/infrastructure_as_software_vs_infrastructure_as/ where basically they make the a distinction between CDK/Pulumi and TF HCL. CDK/Pulumi being infra as software, TF being infra as code and IaSQL being infra as data.